javalib/src/main/scala/java/net/URI.scala

```diff
@@ -601,39 +614,99 @@ object URI {
 
   // Quote helpers
 
-  private val quoteChar: js.Function1[String, String] = { (str: String) =>
-    require(str.length == 1)
+  private def decodeComponent(str: String): String = {
+    // Fast-track, if no encoded components
+    if (str.forall(_ != '%')) str
+    else {
+      val inBuf = CharBuffer.wrap(str)
+      val outBuf = CharBuffer.allocate(inBuf.capacity)
+      val byteBuf = ByteBuffer.allocate(64)
+      val decoder = StandardCharsets.UTF_8.newDecoder
+        .onMalformedInput(CodingErrorAction.REPLACE)
+        .onUnmappableCharacter(CodingErrorAction.REPLACE)
+
+      def decode(endOfInput: Boolean) = {
+        byteBuf.flip()
+        decoder.decode(byteBuf, outBuf, endOfInput)
+        if (endOfInput) {
+          decoder.reset()
+          byteBuf.clear()
+        } else {
+          byteBuf.compact()
+        }
+      }
+
+      def flush() = {
+        if (byteBuf.position > 0) decode(true)
+      }
 
-    val c = str.head.toInt
+      while (inBuf.hasRemaining) {
+        inBuf.get() match {
+          case '%' =>
+            if (!byteBuf.hasRemaining)
+              decode(false)
+
+            // get two chars - they must exist, otherwise the URI would not have
+            // passed syntax checking
+            val hexStr = inBuf.get().toString + inBuf.get().toString
+            val v = Integer.parseInt(hexStr, 16)
+            byteBuf.put(v.toByte)
+
+          case c =>
+            flush()
+            outBuf.put(c)
+        }
+      }
 
-    if (c > 127)
-      throw new URISyntaxException(null, "Only ASCII allowed in URIs")
-    else
-      f"%%$c%02x"
+      flush()
```

Comment:
Alternatively, always call `decode(true)` here (but not in `case c =>`), although that forces one round of decoding.